### DISCUSSION TITLE
umi 项目 全局message 控制台提示[antd: message] Static function can not consume

### DISCUSSION THREAD

**Discussion by desonyuan** [ant-design/ant-design#41855] [General]
### Steps to reproduce

无

### What is expected?

无

### What is actually happening?

无

| Environment | Info |
| --- | --- |
| antd | 5.4.2 |
| React | 18.0.2 |
| System | win10 |
| Browser | chrome |

---
umi 项目 配置axios拦截器中使用了message，应该脱离的项目上下文环境，如何解决这个问题呢？
 控制台提示[antd: message] Static function can not consume context like dynamic theme. Please use 'App' component instead.

**Comment by fat-rab** [comment]
我也遇到了这个问题 ，不知道有没有解决方法

**Reply by Wxh16144** [reply]
暂时有个方案，挂一个空节点，然后用模块儿去引入方法。后续我看看能不能提供一个 demo

**Reply by miss55** [reply]
[可以通过自定义事件来解决全局message问题](https://www.wenyisong.cn/js/antd-message-global/)

**Reply by echoyl** [reply]
妙啊，重写所有message的方法，换一个引用就改完了

**Reply by miss55** [reply]
我把封装代码也贴上了，基本上跟message一模一样
![image](https://github.com/ant-design/ant-design/assets/4410615/33ba7a92-5a1e-43dd-9f02-d04687787cb1)

**Comment by ZYyyh-cool** [comment]
可以提供下demo吗

**Reply by Wxh16144** [reply]
最近有点忙, 不好意思, 大致思路就是直接通过一个全局变量维护

**Reply by Mr-Glin** [reply]
我用的是umi4，能够具体说说吗？我配置之后，  const staticFunctions = App.useApp();会报错，因为我是在src/request.ts中使用的，这个文件用来配置axios的拦截器等

**Reply by Wxh16144** [reply]
https://github.com/ant-design/ant-design/discussions/41855#discussioncomment-5706550

**Reply by Mr-Glin** [reply]
因为用到了umi4框架，不知道在哪里引入EscapeAntd组件从而放入上下文中，在你的demo中，你是放在main.tsx中，但是在umi4中 应该如何配置才能使其生效呢？

**Reply by Wxh16144** [reply]
嗯嗯, 补充一下.
>  const staticFunctions = App.useApp();会报错

需要将组件放在 antd 提供的 App 组件容器下. 否则 useApp 取不到 context

> 但是在umi4中 应该如何配置才能使其生效呢？

没用过 umi, 但是通过文档, 应该可以在 app.tsx 添加运行时? 可以试试. 如果umi 中用了 antd plugin, 可以设置 appConfig 配置启用 App 容器组件既可

**Reply by Mr-Glin** [reply]
我尝试过配置，但是没有效果
![image](https://github.com/ant-design/ant-design/assets/55646980/f3f8507b-9724-4cba-88fa-3a2c7cc1ec40)

**Reply by Wxh16144** [reply]
希望可以帮助到你~
<img width="1512" alt="image" src="https://github.com/ant-design/ant-design/assets/32004925/19b1bb97-b471-4cf0-87ea-ca13aca9d5aa">

**Reply by Mr-Glin** [reply]
非常感谢，但是，我需要在ts中直接使用，而不是在组件中，
![image](https://github.com/ant-design/ant-design/assets/55646980/70a5625b-1c12-4253-ab5a-73f5a402cf32)

**Reply by Wxh16144** [reply]
你这个截图没有问题呀, `EscapeAntd` 必须要挂在 App 容器组件中. 调用才能 App.useApp 将方法放在 global

**Reply by Mr-Glin** [reply]
谢谢，已经可以了。
![image](https://github.com/ant-design/ant-design/assets/55646980/169f72f2-d8ed-4f29-85e4-440b8c5e17e3)
在App.tsx中，增加如下方法
![image](https://github.com/ant-design/ant-design/assets/55646980/cc2cc953-367f-437f-b742-2dd7a92192cc)
于是就可以在ts中直接导入使用了
import { message } from '@/shared/EscapeAntd';
message.success('成功消息');

**Reply by lihongyao** [reply]
你的 antdMessage 哪里来的？

**Reply by minm2015** [reply]
按这种方式确实可以解决报错，可是如何全局设置message的关闭时间了呢，之前可以在app.tsx里面追加message.config({
  duration: 10
})这种方式就可以设置

**Reply by Wxh16144** [reply]
> 如何全局设置message的关闭时间

antd 提供了 `<App />` 组件，可以全局配置，如果你用了 umijs ，可以看看 umijs  antd 的 plugin

**Reply by deepbytelab** [reply]
@Mr-Glin  我也试过这样写，但是 得到是 一个  message 空对象 
![image](https://github.com/ant-design/ant-design/assets/24228407/730b8d0d-1bc3-4452-a8b5-4eb55623530f)

**Reply by noahlam** [reply]
> @Mr-Glin 我也试过这样写，但是 得到是 一个 message 空对象 ![image](https://private-user-images.githubusercontent.com/24228407/318724829-730b8d0d-1bc3-4452-a8b5-4eb55623530f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2NDQ5NTAsIm5iZiI6MTcyODY0NDY1MCwicGF0aCI6Ii8yNDIyODQwNy8zMTg3MjQ4MjktNzMwYjhkMGQtMWJjMy00NDUyLWE4YjUtNGViNTU2MjM1MzBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDExMDQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiOGI1ZWMzNTFlODgxNzFmODlhMWE4Y2Y3NzBhMzg2YTY4OGQxYmZkYjcxOGUwNmI5ZTcyMDc5YmMxYmM3NzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FofVlvwt8yo0bZ4wNqMo3nSts3aOc830YVqJpJ3TBEM)

拿到空对象是因为 EscapeAntd 没有被 ant 的 App 包裹. 需要这样写
```
import { App as AntdApp } from "antd";

<AntdApp>
  <EscapeAntd />
  <App />
</AntdApp>
```

**Comment by xxi-arch** [comment]
+1

**Comment by bairubing-xenera** [comment]
我也是这个问题，看着这个错太难受了

**Comment by Wxh16144** [comment]
我用 vite 项目创建了一个 demo https://github.com/Wxh16144/antd-41855-demo  主要内容在 `src/shared/EscapeAntd.tsx`, 用一个全局变量保持 message 实例, 如果大家有更好的解决方案可以补充一下.

**Reply by XiaoXiaoWu98** [reply]
我试过并没有生效，依然报错，并且不会弹出消息， antd@5.10.2

**Reply by loakliu** [reply]
有用的，参考官网的例子：https://ant-design.antgroup.com/components/app-cn   全局场景，我在next13.4 app route的工程中已经运行可以了

**Reply by nglkz** [reply]
有用，很清晰！

**Reply by XiaoXiaoWu98** [reply]
使用该方法立马使用页面跳转，会仍然不生效，只有做跳转延迟，使用这个方法才生效，页面变更，这个提示就会消失！

**Reply by bitbw** [reply]
Great！ 感谢分享！ 非常棒的解决方案 ~

**Reply by 1127951201** [reply]
不行啊，拿的还是初始的import {
    message as antdMessage,
    notification as antdNotification,
    Modal as antdModal,
    App,
} from "antd";任然会报错

**Comment by big-tutu** [comment]
降了一下版本后没再报错，比如@5.2.1

**Reply by yycang** [reply]
还是这个最方便，眼不见心不烦

**Comment by WuJiY** [comment]
有解决方案了吗？

**Comment by lovewinders** [comment]
+1

**Comment by linxianxi** [comment]
可以试试文档里这样
https://ant.design/components/app-cn#%E5%85%A8%E5%B1%80%E5%9C%BA%E6%99%AFredux-%E5%9C%BA%E6%99%AF

**Reply by Wxh16144** [reply]
> 可以试试文档里这样 https://ant.design/components/app-cn#%E5%85%A8%E5%B1%80%E5%9C%BA%E6%99%AFredux-%E5%9C%BA%E6%99%AF

这个文档比较含糊, 核心成员在讨论并且有计划把他写详细点了🤩

**Comment by nianxi** [comment]
这样是可以的

1. 原app.tsx
> import AntdWrapApp from "./message"
> export function rootContainer(container: React.ReactElement) {
>   return (
>      \<App\>
>         \<AntdWrapApp /\>
>         {container}
>       \</App\>
>    )
> }


2. 新增message.tsx
> import { App } from "antd"
> import type { ModalStaticFunctions } from "antd/es/modal/confirm"
> import type { MessageInstance } from "antd/lib/message/interface"
> import type { NotificationInstance } from "antd/lib/notification/interface"
> 
> let message: MessageInstance
> let notification: NotificationInstance
> let modal: Omit<ModalStaticFunctions, "warn">
> 
> export default () => {
>   const fn = App.useApp()
>   message = fn.message
>   notification = fn.notification
>   modal = fn.modal
>   return null
> }
> export { message, notification, modal }

**Reply by guyuedashu** [reply]
modal组件不支持全局配置

**Reply by guyuedashu** [reply]
export default () => {
  const fn = App.useApp()
  message = fn.message
  notification = fn.notification
  modal = fn.modal

  const _modalWarning = fn.modal.confirm;
  modal.confirm = (data) => {
    _modalWarning(Object.assign({cancelText: '取消', okText: '确定'}, data));
  }
  return null
}
export { message, notification, modal } 
只能这样写了

**Reply by Wxh16144** [reply]
> modal组件不支持全局配置

看文档：https://ant.design/components/app-cn#api

**Reply by terwer** [reply]
这样有问题，暗黑模式还是显示浅色的。

**Reply by echoyl** [reply]
使用全局场景（redux 场景），modal.confirm 的国际化配置无法读取全局的，设置中文显示的还是英文，暗黑模式也还是浅色的。这个有方法吗？

**Reply by sanxin-lin** [reply]
<img width="617" alt="8753d4eba5b73bc9d7cc9e8b0925d37" src="https://github.com/ant-design/ant-design/assets/53200732/46c93963-1233-4e66-8a76-951705a008ef">这样就可以了

**Comment by joesdu** [comment]
不知在[Ant Pro](https://github.com/ant-design/ant-design-pro)应该如何处理?

**Comment by YuShai** [comment]
import { App, message as antdMessage, } from 'antd';
import type { MessageInstance } from 'antd/es/message/interface';

let message: MessageInstance = antdMessage;

function MessageGlobal() {
    const staticFunction = App.useApp();
    message = staticFunction.message;
    return null;
};

export { message, };
export default MessageGlobal;

**Comment by soxft** [comment]
<img width="888" alt="image" src="https://github.com/ant-design/ant-design/assets/42080379/63f189a7-c7c8-4b4d-a5ad-a7d85703d0ec">

try this way

**Reply by soxft** [reply]
不太清楚具体情况唉，但是我这边是可以使用的

**Reply by 928979883** [reply]
我打印了下，发现是我打错了，const [messageApi, contextHolder] = message.useMessage()我用的对象去包裹了，所以没拿到

**Reply by 928979883** [reply]
我就说明明弹窗也是这样解决的，就是拿不到 ，-_-，const [modal, contextHolder] = Modal.useModal()

**Reply by soxft** [reply]
> 我就说明明弹窗也是这样解决的，就是拿不到 ，-_-，const [modal, contextHolder] = Modal.useModal()

啊哈，问题不大，我之前也打错过

**Comment by zooyer** [comment]
export const errorConfig: RequestConfig里面errorHandler的会调用message，这里根本没办法弄，怎么设计的呢。越做越垃圾了怎么，要消费context没问题，但是改接口不考虑兼容性吗？目前项目中就这一个warning，真烦人。
Warning: [antd: message] Static function can not consume context like dynamic theme. Please use 'App' component instead.

**Comment by WuJiY** [comment]
从外面传进来吧，但是request这一层还不知道怎么解决

**Reply by miss55** [reply]
参考这个，通过自定义事件触发即可
[自定义事件](https://github.com/ant-design/ant-design/discussions/41855#discussioncomment-7762212)

**Reply by zooyer** [reply]
你这个不错，解决了。这个本应该官方解决的，自己的antd pro项目跑起来都报错报警，他们自己不考虑吗？

**Comment by davidhsing** [comment]
最合适的方案：
https://github.com/ant-design/pro-components/issues/7000

**Reply by huanghuadong527** [reply]
message.open

**Comment by joebnb** [comment]
感觉这个 API 变更的，不太优雅，上面的开发者都是自己重新封装，没想好的事情不如不做
